docs/source/grammar.rst

```diff
@@ -0,0 +1,242 @@
+    - ``'Hello ${name}!'``
+    - ``"I ${can: ${nest}, ${interpolations}, 'and quotes'}"``
+
+The ``quotedValue`` parser rule is formally defined as:
+
+    .. code-block:: antlr
+
+        quotedValue:
+            (QUOTE_OPEN_SINGLE | QUOTE_OPEN_DOUBLE)
+            text?
+            MATCHING_QUOTE_CLOSE;
+
+
+``listContainer`` and ``dictContainer`` are respectively lists and dictionaries, using a familiar syntax:
+
+    - List examples: ``[]``, ``[1, 2, 3]``, ``[${a}, ${oc.env:B}, c]``
+    - Dict examples: ``{}``, ``{a: 1, b: 2}``, ``{a: ${a}, b: ${oc.env:B}}``
+
+Their corresponding parser rules are:
+
+    .. code-block:: antlr
+
+        listContainer: BRACKET_OPEN sequence? BRACKET_CLOSE;
+        dictContainer: BRACE_OPEN
+                       (dictKeyValuePair (COMMA dictKeyValuePair)*)?
+                       BRACE_CLOSE;
+
+Regarding dictionaries, note that although values can be any ``element``, keys are more
+restricted, and in particular quoted strings and interpolations are currently *not* allowed as
+dictionary keys (see the definition of ``dictKey`` in the `grammar <https://github.com/omry/omegaconf/blob/master/omegaconf/grammar/OmegaConfGrammarParser.g4>`_).
+
+Finally, a ``primitive`` is everything else that is allowed, including in particular (see the `full grammar <https://github.com/omry/omegaconf/blob/master/omegaconf/grammar/OmegaConfGrammarParser.g4>`_
+for details):
+
+    - Unquoted strings (that support only a subset of characters, contrary to quoted ones): ``foo``, ``foo_bar``, ``hello world 123``
+    - Integer numbers: ``123``, ``-5``, ``+1_000_000``
+    - Floating point numbers (with special case-independent keywords for infinity and NaN): ``0.1``, ``1e-3``, ``inf``, ``-INF``, ``nan``
+    - Other special keywords (also case-independent): ``null``, ``true``, ``false``, ``NULL``, ``True``, ``fAlSe``.
+      **IMPORTANT**: ``None`` is *not* a special keyword and will be parsed as an unquoted string, you must
+      use the ``null`` keyword instead (as in YAML).
+    - Interpolations (thus allowing for nested interpolations)
+
+
+Escaped characters
+^^^^^^^^^^^^^^^^^^
+
+Some characters need to be escaped, with varying escaping requirements depending on the situation.
+In general, however, you can use the following rule of thumb:
+*you only need to escape characters that otherwise have a special meaning in the current context*.
+
+.. _escaping-in-interpolation-strings:
+
+Escaping in interpolation strings
++++++++++++++++++++++++++++++++++
+
+In order to define fields whose value is an interpolation-like string, interpolations can be escaped with ``\${``.
+For instance:
+
+.. doctest::
+
+    >>> c = OmegaConf.create({"path": r"\${dir}", "dir": "tmp"})
+    >>> print(c.path)  # does *not* interpolate into the `dir` node
+    ${dir}
+
+Since the ``\`` character has a special meaning in front of ``${``,
+it must itself be escaped with ``\\`` if you actually intend to follow it with an interpolation:
```

Comment:
Not sure I understand what you mean here.
In the above example you use an rstring instead of escaping it.
escaping the `\` is a Python requirement in string literals. Maybe provide an second example without an rstring calling out that this is because `\` has to be escaped in Python strings?

Reply:
Does e71ac8d make it clearer?

Reply:
> Does [e71ac8d](https://github.com/omry/omegaconf/commit/e71ac8db4efb48913565fa198b99296083274d83) make it clearer?

@omry just checking if that commit looks good to you?

Reply:
> > Does [e71ac8d](https://github.com/omry/omegaconf/commit/e71ac8db4efb48913565fa198b99296083274d83) make it clearer?
> 
> @omry just checking if that commit looks good to you?

Actually check this diff which combines the two latest (related) commits: https://github.com/omry/omegaconf/pull/792/files/21f0217c51b562b335eb98af116690fb576bcc6e..29cd49235bfd4c88a733dd0416fb7f5e52176d5f

Reply:
This adjustment is much clearer to me. Thanks